_posts/2021-10-06-jvm-issues-at-comcast.md

```diff
@@ -0,0 +1,219 @@
+---
+layout: post
+title:  "JVM Challenges in Production"
+author: Sajumon Joseph, David Leach, Bryan Aller, Pavan Madhineni, Lavanya Ragothaman, Pratap Moturi, Pádraig O'Sullivan (Starburst)
+excerpt_separator: <!--more-->
+---
+
+At Comcast, we have a large on-premise Trino cluster. It enables us to extract
+insights from data no matter where it resides and prepares the company for a
+more cloud-centric future. Recently, however, we experienced and overcame
+challenges related to the JVM. We wanted to share what we encountered and
+learned in hopes that it might be useful for the Trino community.
+
+<!--more-->
+
+## JIT Recompilations
+
+Some users complained that nightly reports were taking far too long to
+complete. Queries that ran for 6 hours made very little progress.
+
+First, we looked at the queries involved in these nightly reports. We
+noticed that all these queries involved 2 particular tables. In this post,
+we will call them table A and table B.
+
+Our initial suspicion was that there could be an issue with the table data in
+HDFS. Thus, we tried to reproduce the performance problem by using queries that
+performed simple scans against these tables. 
+
+We tried a simple table scan with no filters, range filter on a partitioned
+column, etc.,  ran these queries multiple times and execution times were
+consistent. This ruled out a potential problem with HDFS.
+
+Next, we took a closer look  at the portion of the slow running queries
+involving the A table and came up with the simplest possible query that could
+demonstrate the problem. We discovered that the following query performed as
+expected:
+
+```
+select
+ count(a.c1)
+from
+ hive.schema1.A a, hive.schema2.B da
+where
+ a.day_id = da.date_id
+ and a.day_id BETWEEN '2021-03-22' AND '2021-04-21'
+```
+
+But adding a predicate caused the performance problem to appear:
+
+```
+select
+ count(a.c1)
+from
+ hive.schema1.A a, hive.schema2.date_dim da
+where
+ a.day_id = da.date_id
+ and a.day_id BETWEEN '2021-03-22' AND '2021-04-21'
+ and a.c2 = '4 (Success)'
+```
+
+We narrowed the problem down to the `Scan/Filter/Project` operator using the
+output of `EXPLAIN ANALYZE` from Trino. For the query that performed as
+expected, this stage had the following CPU stats:
+
+```
+CPU: 2.39h, Scheduled: 4.47h, Input: 17434967615 rows (357.47GB)
+```
+
+For the simple version of the query performing poorly, the same stage has the
+following CPU stats:
+
+```
+CPU: 3.73d, Scheduled: 48.01d, Input: 17052985227 rows (413.98GB)
+```
+
+This shows that for roughly the equivalent amount of data, Trino used
+significantly more CPU (3.73 days to 2.39 hours!!). Our next step was to
+determine why that was and exactly what was using additional CPU.
+
+We generated a few jstack and JFR profiles of the Starburst Java process from
```

Comment:
link to jstack maybe 

and what is JFR (spell it out)

And 

of the Trino Java process